File: Minimum-Falling-Path-Sum/MFPS.rs
pub fn min_falling_path_sum(a: Vec<Vec<i32>>) -> i32 {
    use std::cmp::{max, min};
    let leng = a.len();
    let mut a_iter = a.into_iter();
    let mut temp = a_iter.next().unwrap();
    loop {
        if let Some(this_row) = a_iter.next() {
            let mut cache = vec![];
            for (ind, v) in this_row.iter().enumerate() {
                //println!("{}", ind - 1);
                let a = temp[max(ind as i32 - 1, 0) as usize];
                let b = temp[ind];
                let c = temp[min(ind as i32 + 1, leng as i32 - 1) as usize];
                let mm = min(min(a, b), c);
                cache.push(v + mm);
            }
            temp = cache;
        } else {
            break;
        }
    }
    *temp.iter().min().unwrap()
}

fn main() {
    dbg!(min_falling_path_sum(vec![
        vec![1, 2, 3],
        vec![4, 5, 6],
        vec![7, 8, 9]
    ]));
}

File: Minimum-Falling-Path-Sum/MFPS.hs
min_falling_path_sum :: [[Int]] -> Int
min_falling_path_sum (x:xs) = inner_func x xs
  where
    inner_func l []     = minimum l
    inner_func l (x:xs) = inner_func (sum_together l x) xs
    sum_together a b =
      [ (b !! ind) +
      (minimum
         [ (a !! (max (ind - 1) 0))
         , (a !! ind)
         , (a !! (min (ind + 1) (length b - 1)))
         ])
      | ind <- [0 .. ((length b) - 1)]
      ]


main = do
  print $ min_falling_path_sum [[1,2,3],[4,5,6],[7,8,9]]
